posts/2020-11-27-typing-your-actors.adoc

```diff
@@ -0,0 +1,389 @@
+= Typing your actors
+Pedro Ferreira <pedro.ferreira@lunatech.nl>
+v1.0, 2020-11-27
+:title: Typing your actors step by step
+:tags: [akka, migration, akka-typed]
+
+== Background
+
+After some frustration with our project's choice for planning poker solution, I did what any reasonable developer would do. No, I didn't search for a more suitable solution, I decided to write mine and do a blog post about it.
+
+The main idea is having a simple application that involves a bit of the three core Akka libraries:
+
+* Http: REST api for Vue.js frontend
+* Streams: Websocket bi-directional data handling
+* Actors: Rooms management
+
+Initially I chose to use an untyped actor system intentionally, so we can actually talk about something that might be more useful, migrating your old Akka applications to a typed actor system. Why would you want to do that, well for me it is the type safety of not having everything being a function of `Any => Unit` and strong support to avoid anti-patterns (more eloquently explained https://www.youtube.com/watch?v=vF_-28I6yGE[here]), but feel free to search more if this migration is something for your project.
+
+== Overview
+
+Observable behaviour should stay the same after migration, which consists of:
+
+* A stream for each websocket connection with its own wire data https://github.com/lunatech-labs/lunatech-pointingpoker/blob/master/src/main/scala/com/lunatech/pointingpoker/websocket/WSMessage.scala#L10[domain] forwarding all incoming data to the main application controller (Room Manager) as well as providing an interface to forward outgoing data.
+* Room manager is responsible for translating from wire domain to the application https://github.com/lunatech-labs/lunatech-pointingpoker/blob/master/src/main/scala/com/lunatech/pointingpoker/Room.scala#L74-L84[domain] and https://github.com/lunatech-labs/lunatech-pointingpoker/blob/master/src/main/scala/com/lunatech/pointingpoker/RoomManager.scala#L56-L69[route] translated messages to appropriate rooms and manage each room lifecycle.
+* Rooms are representations of each planning poker session, holding all necessary https://github.com/lunatech-labs/lunatech-pointingpoker/blob/master/src/main/scala/com/lunatech/pointingpoker/Room.scala#L13-L15[information] and making sure that each participant is up to date with it.
+
+== Step by step
+
+First things first, we need to replace our old classical actor system with its typed counterpart:
+[source,scala]
+----
```

Comment:
I have no idea if this will work or not with asciidoc code blocks, but I know if you have "github flavor" markdown you can actually put the old code and a new code together and then do "```diff" as the language and then use `+` and `-` where you add or remove stuff. If that does work with asciidoc, it may be a cool way to show what you're changing, if not just ignore this.